c_src/notif_set_tlh_glib.c

```
@@ -0,0 +1,215 @@
+    /* destroy each level2 table */
+    while(g_hash_table_iter_next(&itr1, (void**) &itr1_key, (void**) &lev2_tab)) {
+        g_hash_table_destroy(lev2_tab);
+        /* TODO: Free memory associated to each element in lev2_tab. For now,
+         * we assume this is taken care of by the caller (user) of notif_set.
+         */
+    }
+    g_hash_table_destroy(tab->lev1);
+    free(tab);
+
+    return 1;
+}
+
+int notif_set_insert(struct Notif_set *tab, struct Message *msg)
+{
+    /* IMPORTANT: elements are NOT copied in the
+     * hash tables; only the pointers are stored.
+     */
+    /* msg has three parts:
+     * 1- watch descriptor
+     * 2- length
+     * 3- filename
+     */
+    if(msg == NULL) {
+        return 0;
+    }
+    GHashTable *lev2_tab = g_hash_table_lookup(tab->lev1, 
+                                               &(msg->wd));
+    if(lev2_tab == NULL) {
+        /* we have a new directory wd. Insert to level1 */
+        /* first create a level2 table for this wd (the value) */
+        lev2_tab = g_hash_table_new(g_str_hash, g_str_equal);
+        /* TODO: check for success */
+        g_hash_table_insert(tab->lev1, &(msg->wd), lev2_tab);
+    }
+    
+    /* add filename to the level2 hash table of this wd */
+    /* we use add rather than insert as we only care about the key (filename) */
+    return g_hash_table_add(lev2_tab, msg->name); /* TODO: store len as well */
+}
+
+int notif_set_find(struct Notif_set *tab, struct Message *msg, void *val)
+{
+    (void) val; //To eliminate unused variable warning
+    if(msg == NULL) {
+        return 0;
+    }
+    GHashTable *lev2_tab = g_hash_table_lookup(tab->lev1,
+                                               &(msg->wd));
+    if(lev2_tab != NULL) {
+        if(g_hash_table_contains(lev2_tab, msg->name)) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+int notif_set_del(struct Notif_set *tab, struct Message *msg)
+{
+    if(msg == NULL) {
+        return 0;
```

I would consider passing a `NULL` as the key to delete to be a programmer error and should therefor `abort`. In other words this is a breach of the `notif_set_del` contract and therefore should fail hard.

---

Reading the wrong file... nevermind...